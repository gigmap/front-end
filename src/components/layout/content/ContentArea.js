// @flow

import React, {Suspense} from 'react';
import {connect} from 'react-redux';
import LoadingOverlay from '../../common/loading-overlay/LoadingOverlay';
import Loading from '../../common/loading/Loading';
import {Alert, Button} from 'antd';
import {isAuthenticated} from '../../../store/reducers/user/selectors';
import {load} from '../../../store/actions/data';
import styles from './ContentArea.module.less';

const ConcertMap = React.lazy(() => import('../../concerts/map/ConcertMap'));
const WelcomePage = React.lazy(() => import('../../welcome-page/WelcomePage'));

type ContentAreaProps = {
  authenticated: boolean,
  finished: boolean,
  error: string,
  loading: boolean,
  load: Function
};

// TODO: rework error detecting and rendering
const renderError = (error: string, load: Function) => {
  const notice = error ?
    <Alert type='error' showIcon
           message={
             <div>An error happened during concert loading:<br/>{error}</div>
           }
    /> :
    <Alert type='warning' showIcon
           message='Somehow concerts are not loaded.'/>;

  const button =
    <Button type='primary' className={styles.button} onClick={() => load(true)}>
      Try Again
    </Button>;

  return <div className={styles.errorWrapper}>
    <div>
      {notice}
      {button}
    </div>
  </div>;
};

function renderContent(props: ContentAreaProps) {
  const {authenticated, finished, error, load, loading} = props;

  if (!authenticated) {
    return <Suspense fallback={<Loading/>}>
      <WelcomePage/>
    </Suspense>;
  }

  if (!loading && (!finished || error)) {
    return renderError(error, load);
  }

  return (
    <Suspense fallback={<Loading/>}>
      <ConcertMap/>
    </Suspense>
  );
}

export const ContentArea = (props: ContentAreaProps) => {
  const {loading} = props;

  return (
    <>
      {loading && <LoadingOverlay/>}
      {renderContent(props)}
    </>
  );
};

const mapStateToProps = (state) =>
  ({
    loading: state.data.loading,
    finished: state.data.finished,
    error: state.data.error,
    authenticated: isAuthenticated(state)
  });

const mapDispatchToProps = {load};

export const ConnectedContentArea =
  connect(mapStateToProps, mapDispatchToProps)(ContentArea);