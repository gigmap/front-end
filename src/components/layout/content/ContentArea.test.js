import React from 'react';
import {shallow} from 'enzyme';
import {ContentArea} from './ContentArea';

describe('App', function () {
  function setup(propsOverrides) {
    const props = {
      loading: false,
      authenticated: false,
      ...propsOverrides
    };

    const wrapper = shallow(<ContentArea {...props} />);
    return {wrapper};
  }

  describe('snapshots', function () {
    [
      {loading: false, authenticated: false, finished: false},
      {loading: false, authenticated: true, finished: true},
      {loading: true, authenticated: false, finished: false},
      {loading: true, authenticated: true, finished: false}
    ].forEach(props => {
      it(`should render without crashing with props ${JSON.stringify(props)}`, () => {
        const {wrapper} = setup(props);
        expect(wrapper).toMatchSnapshot();
      });
    })
  });

  // TODO: mount to test lazy components
});