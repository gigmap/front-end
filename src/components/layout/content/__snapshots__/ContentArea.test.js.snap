// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App snapshots should render without crashing with props {"loading":false,"authenticated":false} 1`] = `
<Fragment>
  <Suspense
    fallback={<Loading />}
  >
    <lazy />
  </Suspense>
</Fragment>
`;

exports[`App snapshots should render without crashing with props {"loading":false,"authenticated":true} 1`] = `
<Fragment>
  <Suspense
    fallback={<Loading />}
  >
    <lazy />
  </Suspense>
</Fragment>
`;

exports[`App snapshots should render without crashing with props {"loading":true,"authenticated":false} 1`] = `
<Fragment>
  <LoadingOverlay />
  <Suspense
    fallback={<Loading />}
  >
    <lazy />
  </Suspense>
</Fragment>
`;

exports[`App snapshots should render without crashing with props {"loading":true,"authenticated":true} 1`] = `
<Fragment>
  <LoadingOverlay />
  <Suspense
    fallback={<Loading />}
  >
    <lazy />
  </Suspense>
</Fragment>
`;
